pkg/scheduler/queue/tenant_querier_assignment.go

```diff
@@ -358,3 +311,160 @@ func (tqa *tenantQuerierAssignments) shuffleTenantQueriers(tenantID TenantID, sc
 	tqa.tenantQuerierIDs[tenantID] = querierIDSet
 	return true
 }
+
+// dequeueGetNode chooses the next node to dequeue from based on tenantIDOrder and tenantOrderIndex, which are
+// shared across all nodes to maintain an O(n) (where n = # tenants) time-to-dequeue for each tenant.
+// If tenant order were maintained by individual nodes, we would end up with O(mn) (where m = # query components)
+// time-to-dequeue for a given tenant.
+//
+// tenantOrderIndex is incremented, checks if the tenant at that index is a child of the current node (it may not be,
+// e.g., in the case that a tenant has queries queued for one query component but not others), and if so, returns
+// the tenant node if currentQuerier can handle queries for that tenant.
+//
+// Note that because we use the shared  tenantIDOrder and tenantOrderIndex to manage the queue, we functionally
+// ignore each Node's individual queueOrder and queuePosition.
+func (tqa *tenantQuerierAssignments) dequeueSelectNode(node *Node) (*Node, bool) {
+	// can't get a tenant if no querier set
+	if tqa.currentQuerier == nil {
+		return nil, true
+	}
+
+	checkedAllNodes := node.childrenChecked == len(node.queueMap)+1 // must check local queue as well
+
+	// advance queue position for dequeue
+	tqa.tenantOrderIndex++
+	if tqa.tenantOrderIndex >= len(tqa.tenantIDOrder) {
+		tqa.tenantOrderIndex = localQueueIndex
+	}
+
+	// no children or local queue reached
+	if len(node.queueMap) == 0 || tqa.tenantOrderIndex == localQueueIndex {
+		return node, checkedAllNodes
+	}
+
+	checkIndex := tqa.tenantOrderIndex
+
+	// iterate through the tenant order until we find a tenant that is assigned to the current querier, or
+	// have checked the entire tenantIDOrder, whichever comes first
+	for iters := 0; iters < len(tqa.tenantIDOrder); iters++ {
+		if checkIndex >= len(tqa.tenantIDOrder) {
+			// do not use modulo to wrap this index; tenantOrderIndex is provided from an outer process
+			// which does not know if the tenant list has changed since the last dequeue
+			// wrapping with modulo after shrinking the list could cause us to skip tenants
+			checkIndex = 0
+		}
+		tenantID := tqa.tenantIDOrder[checkIndex]
+		tenantName := string(tenantID)
+
+		if _, ok := node.queueMap[tenantName]; !ok {
+			// tenant not in _this_ node's children, move on
+			checkIndex++
+			continue
+		}
+
+		// childrenChecked is used to determine when we can stop looking for a dequeueable node. At this point,
+		// tenantID _must_ exist in the node's children, even if currentQuerier isn't assigned to it, so we increment
+		// childrenChecked regardless of whether currentQuerier exists in tenantQuerierIDs for tenantID.
+		node.childrenChecked++
```

Comment:
This is another spot where multiple objects are responsible for the same state (Node + TQA), which makes the tree much harder to understand

Reply:
This is a bug! I moved this increment into dequeue based on some earlier PR feedback, but missed this. Thanks for catching it!

I think this ended up passing tests because we update `checkedAllNodes` in `dequeue()` before we actually go through with our dequeue operations, and because we enforce that every child node _actually has_ something and never use `childrenChecked` to set queue position, we end up dequeuing the relevant thing before we actually use `checkedAllNodes` to break out of the loop.

I'm not _sure_ how to write a test for this in this case, open to thoughts.

Reply:
I am still on the side of moving  `childrenChecked`, `queuePosition` and `queueOrder` from the nodes to the algo completely, as they are state of the algorithm's selection cycle, not the tree itself.

For instance if we were have an algorithm that is strictly least-load, `queuePosition` and `queueOrder` won't be utilized at all.

Reply:
I think there's a reasonable argument for moving `queueOrder` and `queuePosition` out of the nodes, but I'm not sure I understand the argument for moving `childrenChecked` out of the node. It's currently only updated by the tree queue, and moving it out would require algorithm implementers to remember to handle it.